besu/src/test/java/org/hyperledger/besu/controller/QbftBesuControllerBuilderTest.java

```diff
@@ -0,0 +1,169 @@
+
+@RunWith(MockitoJUnitRunner.class)
+public class QbftBesuControllerBuilderTest {
+
+  private BesuControllerBuilder qbftBesuControllerBuilder;
+
+  @Mock private GenesisConfigFile genesisConfigFile;
+  @Mock private GenesisConfigOptions genesisConfigOptions;
+  @Mock private SynchronizerConfiguration synchronizerConfiguration;
+  @Mock private EthProtocolConfiguration ethProtocolConfiguration;
+  @Mock private MiningParameters miningParameters;
+  @Mock private ObservableMetricsSystem observableMetricsSystem;
+  @Mock private PrivacyParameters privacyParameters;
+  @Mock private Clock clock;
+  @Mock private TransactionPoolConfiguration poolConfiguration;
+  @Mock private StorageProvider storageProvider;
+  @Mock private GasLimitCalculator gasLimitCalculator;
+  @Mock private WorldStateStorage worldStateStorage;
+  @Mock private WorldStatePreimageStorage worldStatePreimageStorage;
+  private static final BigInteger networkId = BigInteger.ONE;
+  private static final NodeKey nodeKey = NodeKeyUtils.generate();
+
+  @Rule public final TemporaryFolder tempDirRule = new TemporaryFolder();
+
+  @Before
+  public void setup() {
+    // besu controller setup
+    when(genesisConfigFile.getConfigOptions()).thenReturn(genesisConfigOptions);
+    when(genesisConfigFile.getParentHash()).thenReturn(Hash.ZERO.toHexString());
+    when(genesisConfigFile.getDifficulty()).thenReturn(Bytes.of(0).toHexString());
+    when(genesisConfigFile.getExtraData()).thenReturn(Bytes.EMPTY.toHexString());
+    when(genesisConfigFile.getMixHash()).thenReturn(Hash.ZERO.toHexString());
+    when(genesisConfigFile.getNonce()).thenReturn(Long.toHexString(1));
+    when(genesisConfigFile.getConfigOptions(any())).thenReturn(genesisConfigOptions);
+    when(storageProvider.createBlockchainStorage(any()))
+        .thenReturn(
+            new KeyValueStoragePrefixedKeyBlockchainStorage(
+                new InMemoryKeyValueStorage(), new MainnetBlockHeaderFunctions()));
+    when(storageProvider.createWorldStateStorage(DataStorageFormat.FOREST))
+        .thenReturn(worldStateStorage);
+    when(worldStateStorage.isWorldStateAvailable(any(), any())).thenReturn(true);
+    when(worldStateStorage.updater()).thenReturn(mock(WorldStateStorage.Updater.class));
+    when(worldStatePreimageStorage.updater())
+        .thenReturn(mock(WorldStatePreimageStorage.Updater.class));
+    when(storageProvider.createWorldStatePreimageStorage()).thenReturn(worldStatePreimageStorage);
+    when(synchronizerConfiguration.getDownloaderParallelism()).thenReturn(1);
+    when(synchronizerConfiguration.getTransactionsParallelism()).thenReturn(1);
+    when(synchronizerConfiguration.getComputationParallelism()).thenReturn(1);
+    when(observableMetricsSystem.createLabelledCounter(
+            any(), anyString(), anyString(), anyString()))
+        .thenReturn(labels -> null);
+    when(synchronizerConfiguration.getBlockPropagationRange()).thenReturn(Range.closed(1L, 2L));
```

Comment:
If I could review my own code...I think there's a question a question of test value vs potential brittleness here. 
If future changes to the BesuControllerBuilder warrant more mocking changes then is it the test worth the hassle?
Those changes would also break BesuControllerBuilderTest, so someone would have to fix that too.

I'm not sure how often BesuControllerBuilder will change in a way that causes this test to fail.

Reply:
Think overall there is more benefit to having the unit tests. But  like you said they could end up being brittle and also if either of these tests failed then Besu build would fail anyway as Qbft would never start up in the QBFT ATs. But still like the early warning and specific failures from the unit tests.